ouroboros-consensus/src/ouroboros-consensus/Ouroboros/Consensus/Util/LeakyBucket.hs

```diff
@@ -126,62 +191,77 @@ runAgainstBucket ::
   (MonadDelay m, MonadAsync m, MonadFork m, MonadMask m, NoThunks (m ())) =>
   Config m ->
   (Handlers m -> m a) ->
-  m (State (Config m), a)
+  m (State m, a)
 runAgainstBucket config action = do
+  runThreadVar <- atomically newEmptyTMVar
   bucket <- init config
-  tid <- myThreadId
-  killThreadVar <- newTVarIO Nothing
-  finally
-    ( do
-        startThread killThreadVar bucket tid
-        result <-
-          action $
-            Handlers
-              { fill = (snd <$>) . snapshotFill bucket,
-                setPaused = setPaused bucket,
-                updateConfig = updateConfig killThreadVar bucket tid
-              }
-        state <- snapshot bucket
-        pure (state, result)
-    )
-    (stopThread killThreadVar)
+  bracket
+    (spawnThreadAndMaybeStart runThreadVar bucket)
+    killThread
```

Comment:
Couldn't we now use `withAsync` or even `withRegistry` here instead of the "low-level" operations `async` and `uninterruptibleCancel`?

Reply:
Absolutely! Done in 09665f2de91dc3e8ad6e7439b0263fa0e7c455ae.